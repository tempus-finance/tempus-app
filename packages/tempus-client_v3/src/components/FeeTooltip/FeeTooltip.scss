.tc__fee-tooltip__content {
  width: 294px;
  padding: 16px 16px 0 16px;
  text-align: left;
  box-sizing: border-box;

  .tc__fee-tooltip__fees {
    margin: 12px 0 20px 0;
  }

  .tc__fee-tooltip__fee-row {
    display: flex;
    justify-content: space-between;
    margin-bottom: 8px;
  }

  hr {
    margin: 0 -16px;
    border: none;
    border-top: 1px solid var(--tooltipBorder);
  }

  .tc__fee-tooltip__info {
    margin: 8px 0;
    display: block;
    text-align: center;
  }
}
