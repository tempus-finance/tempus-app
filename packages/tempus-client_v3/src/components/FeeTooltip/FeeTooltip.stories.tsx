import { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import FeeTooltip from './FeeTooltip';

export default {
  title: 'FeeTooltip',
  component: FeeTooltip,
  argTypes: {},
} as ComponentMeta<typeof FeeTooltip>;

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '50px',
};

const Template: ComponentStory<typeof FeeTooltip> = () => {
  const [tooltipOpen, setTooltipOpen] = useState(false);

  const toggleTooltip = () => {
    setTooltipOpen(prevState => !prevState);
  };

  return (
    <div style={style}>
      <button type="button" onClick={toggleTooltip} style={{ position: 'relative' }}>
        Click me!
        {/* Tooltip should be placed inside it's anchor element - and anchor element must have relative position */}
        <FeeTooltip open={tooltipOpen} fees={{ swap: 0.002 }} />
      </button>
    </div>
  );
};

export const Primary = Template.bind({});
