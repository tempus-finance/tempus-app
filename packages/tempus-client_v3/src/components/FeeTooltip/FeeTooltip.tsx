import { FC, memo, useMemo } from 'react';
import { DecimalUtils, Numberish } from 'tempus-core-services';
import { TooltipWrapper, Typography } from '../shared';
import './FeeTooltip.scss';

export interface FeeTooltipFees {
  deposit?: Numberish;
  redemption?: Numberish;
  earlyRedemption?: Numberish;
  swap?: Numberish;
}

interface FeeTooltipProps {
  fees: FeeTooltipFees;
}

const FeeTooltip: FC<FeeTooltipProps> = props => {
  const { fees, children } = props;
  const tooltipContent = useMemo(
    () => (
      <div className="tc__fee-tooltip__content">
        <Typography variant="body-primary" weight="bold">
          Fee
        </Typography>
        <div className="tc__fee-tooltip__fees">
          <div className="tc__fee-tooltip__fee-row">
            <Typography variant="body-primary" color="text-disabled">
              Deposit
            </Typography>
            <Typography type="mono" variant="body-primary" color="text-primary">
              {DecimalUtils.formatPercentage(fees.deposit ?? 0)}
            </Typography>
          </div>
          <div className="tc__fee-tooltip__fee-row">
            <Typography variant="body-primary" color="text-disabled">
              Redemption
            </Typography>
            <Typography type="mono" variant="body-primary" color="text-primary">
              {DecimalUtils.formatPercentage(fees.redemption ?? 0)}
            </Typography>
          </div>
          <div className="tc__fee-tooltip__fee-row">
            <Typography variant="body-primary" color="text-disabled">
              Early Redemption
            </Typography>
            <Typography type="mono" variant="body-primary" color="text-primary">
              {DecimalUtils.formatPercentage(fees.earlyRedemption ?? 0)}
            </Typography>
          </div>
          <div className="tc__fee-tooltip__fee-row">
            <Typography variant="body-primary" color="text-disabled">
              Swap
            </Typography>
            <Typography type="mono" variant="body-primary" color="text-primary">
              {DecimalUtils.formatPercentage(fees.swap ?? 0)}
            </Typography>
          </div>
        </div>
        <hr />
        <span className="tc__fee-tooltip__info">
          <Typography variant="body-secondary" color="text-disabled">
            Swap fees will go to Liquidity providers
          </Typography>
        </span>
      </div>
    ),
    [fees.deposit, fees.earlyRedemption, fees.redemption, fees.swap],
  );

  return <TooltipWrapper tooltipContent={tooltipContent}>{children}</TooltipWrapper>;
};

export default memo(FeeTooltip) as FC<FeeTooltipProps>;
