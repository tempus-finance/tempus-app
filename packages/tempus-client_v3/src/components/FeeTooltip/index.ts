import FeeTooltip from './FeeTooltip';

export default FeeTooltip;
