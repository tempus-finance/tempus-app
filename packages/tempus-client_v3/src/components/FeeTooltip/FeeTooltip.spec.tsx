import { fireEvent, render } from '@testing-library/react';
import FeeTooltip, { FeeTooltipFees, FeeTooltipProps } from './FeeTooltip';

const fees: FeeTooltipFees = {
  deposit: 0,
  redemption: 0,
  earlyRedemption: 0.003,
  swap: 0.002,
};

const defaultProps: FeeTooltipProps = {
  fees,
};

const subject = (props: FeeTooltipProps) => render(<FeeTooltip {...props}>Click</FeeTooltip>);

describe('FeeTooltip', () => {
  it('renders a fee tooltip', () => {
    const { container, getByText } = subject(defaultProps);

    const actualButton = getByText('Click');

    expect(actualButton).not.toBeNull();

    fireEvent.click(actualButton);

    const tooltip = container.querySelector('.tc__tooltip');

    expect(tooltip).not.toBeNull();
    expect(tooltip).toMatchSnapshot();
  });

  it('renders a fee tooltip with all fees set to 0', () => {
    const { container, getByText } = subject({ fees: {} });

    const actualButton = getByText('Click');

    expect(actualButton).not.toBeNull();

    fireEvent.click(actualButton);

    const tooltip = container.querySelector('.tc__tooltip');

    expect(tooltip).not.toBeNull();
    expect(tooltip).toMatchSnapshot();
  });

  it('does not render a fee tooltip if it is not opened', () => {
    const { container, getByText } = subject(defaultProps);

    const actualButton = getByText('Click');

    expect(actualButton).not.toBeNull();

    const tooltip = container.querySelector('.tc__tooltip');

    expect(tooltip).toBeNull();
  });
});
