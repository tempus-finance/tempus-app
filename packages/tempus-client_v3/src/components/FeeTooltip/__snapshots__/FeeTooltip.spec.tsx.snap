// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeTooltip renders a fee tooltip 1`] = `
<div
  class="tc__tooltip tc__tooltip__bottom-right"
>
  <div
    class="tc__fee-tooltip__content"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Fee
    </div>
    <div
      class="tc__fee-tooltip__fees"
    >
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Deposit
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Redemption
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Early Redemption
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.30%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Swap
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.20%
        </div>
      </div>
    </div>
    <hr />
    <span
      class="tc__fee-tooltip__info"
    >
      <div
        style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        Swap fees will go to Liquidity providers
      </div>
    </span>
  </div>
</div>
`;

exports[`FeeTooltip renders a fee tooltip with all fees set to 0 1`] = `
<div
  class="tc__tooltip tc__tooltip__bottom-right"
>
  <div
    class="tc__fee-tooltip__content"
  >
    <div
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      Fee
    </div>
    <div
      class="tc__fee-tooltip__fees"
    >
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Deposit
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Redemption
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Early Redemption
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
      <div
        class="tc__fee-tooltip__fee-row"
      >
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Swap
        </div>
        <div
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          0.00%
        </div>
      </div>
    </div>
    <hr />
    <span
      class="tc__fee-tooltip__info"
    >
      <div
        style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(122, 122, 122); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
      >
        Swap fees will go to Liquidity providers
      </div>
    </span>
  </div>
</div>
`;
